internal/tunnel/argo.go

```diff
@@ -110,32 +116,89 @@ func (mgr *ArgoTunnelManager) Active() bool {
 
 func (mgr *ArgoTunnelManager) Start(serviceURL string) error {
 
+	mgr.tunnelConfig.Logger.WithFields(log.Fields{
+		"service":  mgr.Config().ServiceName,
+		"hostname": mgr.Config().ExternalHostname,
+	}).Infof("starting tunnel")
+
 	if serviceURL == "" {
 		return fmt.Errorf("Cannot start tunnel for %s with empty url", mgr.Config().ServiceName)
 	}
 	mgr.tunnelConfig.OriginUrl = serviceURL
 
 	placeHolderOnlyConnectedSignal := make(chan struct{})
 	mgr.stopCh = make(chan struct{})
+	mgr.errCh = make(chan error)
 
+	// error handler
+	go func() {
+		mgr.tunnelConfig.Logger.WithFields(log.Fields{
+			"service":  mgr.Config().ServiceName,
+			"hostname": mgr.Config().ExternalHostname,
+		}).Infof("Starting error receiver")
+		for {
+			select {
+			case err, open := <-mgr.errCh:
+				if !open {
+					return
+				}
+				mgr.tunnelConfig.Logger.WithFields(log.Fields{
+					"service":  mgr.Config().ServiceName,
+					"hostname": mgr.Config().ExternalHostname,
+				}).Errorf("Received error (%s) %v", reflect.TypeOf(err), err)
+
+				// the errors are all un-exported from origin package :(
+				// hard to figure out when to retry then
+				switch err.(type) {
+
+				case argoTunnelStopError:
+					mgr.tunnelConfig.Logger.WithFields(log.Fields{
+						"service":  mgr.Config().ServiceName,
+						"hostname": mgr.Config().ExternalHostname,
+					}).Errorf("Stopping tunnel")
+					return
+
+				default:
+					mgr.tunnelConfig.Logger.WithFields(log.Fields{
+						"service":  mgr.Config().ServiceName,
+						"hostname": mgr.Config().ExternalHostname,
+					}).Errorf("Retrying tunnel")
+					mgr.Start(serviceURL)
+				}
+			}
+		}
+	}()
+
+	// tunnel start
 	go func() {
 		mgr.errCh <- origin.StartTunnelDaemon(mgr.tunnelConfig, mgr.stopCh, placeHolderOnlyConnectedSignal)
 	}()
+
 	return nil
 }
 
 func (mgr *ArgoTunnelManager) Stop() error {
+	mgr.tunnelConfig.Logger.WithFields(log.Fields{
+		"service":  mgr.Config().ServiceName,
+		"hostname": mgr.Config().ExternalHostname,
+	}).Infof("stopping tunnel")
+
 	if mgr.stopCh == nil {
 		return fmt.Errorf("tunnel %s already stopped", mgr.id)
 	}
+	mgr.errCh <- argoTunnelStopError{}
```

Comment:
stop should also close the channel, feels like it should be (I may need to build myself a test program to verify)
```
close(mgr.stopCh)
close(mgr.errCh)
mgr.tunnelConfig.OriginUrl = ""
mgr.stopCh = nil
```

I also don't think you want to send `argoTunnelStopError` down the channel.  It looks like it will terminate the channel consumer (error handler), `argoTunnelStopError` triggers function return prior to attempting to stop the controller (e.g. this looks sufficient to capture errors while running and retry start, but it doesn't look sufficient to trap errors during shutdown and retry shutdown ... arguably the re-start flow is the more important flow, just point it out).

Reply:
agree that the lifecycle of the error and stop channels need a little more thought, perhaps shouldn't belong to the manager object at all.